### ISSUE TITLE
Wrong scale factor with convertUnits under Windows

### ISSUE THREAD

**Issue by adeas31** [OpenModelica/OpenModelica#11397]
### Description

The `convertUnits` API returns the wrong scale factor in Windows.

### Steps to Reproduce

Run,

```
convertUnits("m", "mm");
getErrorString();
```

### Expected Behavior

The value of `0.001` should be returned instead of `0.0010000000000000002`.

### Additional Context

@perost tested this on Linux and it works fine.
I debugged a bit more and it seems the value is changed when returning from C to MetaModelica,
`UnitParserExt_omc.cpp:62` if I print `scaleFactor` here I get `0.001`
`UnitAbsynBuilder.mo:1561` if I print `scaleFactor `here I get `0.0010000000000000002`

Related to #8860

**Comment by casella** [comment]
I guess UnitAbsynBuilder doesn't use Ryu yet...

**Comment by phannebohm** [comment]
I just checked and the binary representations of `0.001` and `0.0010000000000000002` are actually different in the last binary digit. So it's not ryu's fault but the computation is probably not exact.

**Comment by casella** [comment]
> I just checked and the binary representations of `0.001` and `0.0010000000000000002` are actually different in the last binary digit. So it's not ryu's fault but the computation is probably not exact.

@phannebohm I'm not sure I understand how this can be the case. The conversion factor between m an mm is 1/1000 or 0.001. Howe could it be that we get a result that is different bitwise from 0.001?

**Comment by perost** [comment]
> > I just checked and the binary representations of `0.001` and `0.0010000000000000002` are actually different in the last binary digit. So it's not ryu's fault but the computation is probably not exact.
> 
> @phannebohm I'm not sure I understand how this can be the case. The conversion factor between m an mm is 1/1000 or 0.001. Howe could it be that we get a result that is different bitwise from 0.001?

`0.001` can't be represented exactly in double-precision floating point, the closest representation is `0.0010000000000000000208...`. Note however that this is a different number to the `0.0010000000000000002` that we get, which has one zero less between the 1 and 2.

So if the computation does involve `0.001` it's not going to be exact. But that doesn't really explain why we get different results on Windows and Linux though...

**Comment by casella** [comment]
Where do I find the source code of `convertUnits()`?

**Comment by perost** [comment]
> Where do I find the source code of `convertUnits()`?

`convertUnits` is defined as:
https://github.com/OpenModelica/OpenModelica/blob/fd4e05a11b3405ef38fa96ffaf0767529752009a/OMCompiler/Compiler/Script/CevalScriptBackend.mo#L782-L796

`UnitAbsynBuilder.str2unitWithScaleFactor` calls the external C-function `UnitParserExt_str2unit` from [UnitParserExt_omc.cpp](https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/Compiler/runtime/UnitParserExt_omc.cpp) (which uses classes from other unitparser files in the same folder).

**Comment by casella** [comment]
I guess then scaleFactor2=0.001 and scaleFactor1=1. How does realDiv work?

**Comment by perost** [comment]
> I guess then scaleFactor2=0.001 and scaleFactor1=1. How does realDiv work?

It's just old MetaModelica for regular division, `scaleFactor2 / scaleFactor1`.

**Comment by casella** [comment]
I guess division by 1 should not change the first argument. Can you track with gdb why scaleFactor2 is not 0.001 but something one bit different?

**Comment by adeas31** [comment]
@casella I don't think the issue is here. As I wrote in the description,

> @perost tested this on Linux and it works fine.
> I debugged a bit more and it seems the value is changed when returning from C to MetaModelica,
> UnitParserExt_omc.cpp:62 if I print scaleFactor here I get 0.001
> UnitAbsynBuilder.mo:1561 if I print scaleFactor here I get 0.0010000000000000002

The value is changed when returning from C to MetaModelica.

**Comment by perost** [comment]
> I guess division by 1 should not change the first argument. Can you track with gdb why scaleFactor2 is not 0.001 but something one bit different?

I've now thoroughly checked what's happening, and for me on Linux everything is fine. The value is the closest representation of 0.001 all the way from the external unit parsing to being printed by `realString` in MetaModelica.

So the only thing I can think of is that the floating point operations are done differently on Windows somehow. Maybe someone using Windows could try compiling OMC with `-msse2 -mfpmath=sse` and see if that solves the issue? SSE should be the default on 64-bit platforms, but who knows with MinGW.

**Comment by casella** [comment]
I find it a bit odd that the result of floating point operations depend on the operating system. But anyway, from @adeas31 comment
> UnitParserExt_omc.cpp:62 if I print scaleFactor here I get 0.001
> UnitAbsynBuilder.mo:1561 if I print scaleFactor here I get 0.0010000000000000002
I understand the result of the division is correct, and it is corrupted when passing from C to metamodelica. @adeas31 I'm not sure how to do you print those values, are you using Ryu for both of them? Could you also get the binary dump of the double number in both cases and check if the bits are unchanged?

**Comment by adrpo** [comment]
Hm, could it be clang vs gcc? What did you compile on Linux with?
I added the sse flags to everything and the result is the same. I will continue debugging.

**Comment by perost** [comment]
> Hm, could it be clang vs gcc? What did you compile on Linux with? I added the sse flags to everything and the result is the same. I will continue debugging.

I think I tried it before but just to be sure I tested it again with both GCC vs Clang and CMake vs autotools, and it makes no difference what I compile with.

**Comment by adrpo** [comment]
`*scaleFactor = unit.scaleFactor.toReal() * pow(10,unit.prefixExpo.toReal());`

this is what on windows gets me:
```
(gdb) p *scaleFactor
$4 = 0.0010000000000000002
```

**Comment by adrpo** [comment]
So the problem is actually in `OMCompiler\Compiler\runtime\UnitParserExt_omc.cpp`. I can now check what the ... is going on.

**Comment by adrpo** [comment]
```
(gdb) p  unit.scaleFactor
$6 = {_vptr.Rational = 0x702b3170 <vtable for Rational+16>, num = 1,
  denom = 1}
(gdb) p unit.prefixExpo
$7 = {_vptr.Rational = 0x702b3170 <vtable for Rational+16>, num = -3,
  denom = 1}
```
found it:

```
(gdb) p 1*pow(10, -3)
$9 = 0.0010000000000000002
(gdb) p pow(10, -3)
$10 = 0.0010000000000000002
```

so `pow `gives us that fabulos number :)

**Comment by adrpo** [comment]
I checked on Linux and windows with this program 
#include <math.h>

using namespace std;

```c
int main()
{
    double x = pow(10, -3);
    printf("%20.20g\n", x);
    return 0;
}
```
and I get:
```
adrpo33@ida-0030 MINGW64 /c/home/adrpo33/dev/OMTesting
# ./pow
0.0010000000000000000208
```

so this but might actually be inside ryu displaying the same number differently on Linux vs Windows.

**Comment by phannebohm** [comment]
I tried a few compilers at: https://godbolt.org/z/Kff1M7KKW
It seems MinGW clang reproduces the error while others don't.

**Comment by perost** [comment]
As mentioned earlier in the issue 0.0010000000000000000208... is the closest double-precision representation of 0.001, so it just means that `pow(10, -3)` is calculated as exactly as is possible. It's also not the number that's actually printed in the end:
```
pow(10, -3):            0.0010000000000000000208
result of convertUnits: 0.0010000000000000002
```

**Comment by adrpo** [comment]
Strangely gdb shows it as Ryu does:
```
(gdb) p 1*pow(10, -3)
$9 =  0.0010000000000000002
Pow: 0.0010000000000000000208
```

**Comment by phannebohm** [comment]
@adrpo what compiler are you using? I guess if it's MinGW clang 16.0.0 or lower, we should update to a newer version if possible.

**Comment by adrpo** [comment]
I will have a meeting with @AnHeuermann tomorrow to plan the switch to a new OMDev based on UCRT. I will check if that helps.

**Comment by adrpo** [comment]
currently we have this in OMDev:
```
adrpo33@ida-0030 MINGW64 /c/home/adrpo33/dev/OpenModelica
# clang --version
clang version 11.0.0 (https://github.com/msys2/MINGW-packages c0083b9edc111315c868f9c53dd60aaa78d1669b)
Target: x86_64-w64-windows-gnu
Thread model: posix
InstalledDir: C:\OMDev\tools\msys\mingw64\bin

adrpo33@ida-0030 MINGW64 /c/home/adrpo33/dev/OpenModelica
# gcc --version
gcc.exe (Rev5, Built by MSYS2 project) 10.2.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

**Comment by adrpo** [comment]
```
model pow
  parameter Real a = 10^(-3);
end pow;
```

windows master with ryu:
```
adrpo33@ida-0030 MINGW64 /c/home/adrpo33/dev/OMTesting
# omc pow.mo
class pow
  parameter Real a = 0.0010000000000000002;
end pow;
```
Linux master:
```
adrpo33@ida-0030:/mnt/c/home/adrpo33/dev/OMTesting$ omc pow.mo
class pow
  parameter Real a = 0.001;
end pow;
```

windows without ryu (v1.22.0)
```
adrpo33@ida-0030 MINGW64 /c/home/adrpo33/dev/OMTesting
# /e/Program\ Files/OpenModelica1.22.0-dev.beta.6-64bit/bin/omc pow.mo
class pow
  parameter Real a = 0.001;
end pow;
```

**Comment by phannebohm** [comment]
Instead of computing `pow(10,unit.prefixExpo.toReal())`, you could use a local lookup table if there are only a few possible exponents for `unit.prefixExpo`. That would solve the precision issue and might even be faster than calling `pow`.

**Comment by perost** [comment]
> Instead of computing `pow(10,unit.prefixExpo.toReal())`, you could use a local lookup table if there are only a few possible exponents for `unit.prefixExpo`. That would solve the precision issue and might even be faster than calling `pow`.

I think `pow` is just a red herring in this case, since as mentioned it already seems to be computed as exactly as possible. Can anyone check how omc on Windows prints this?
```Modelica
model pow
  parameter Real a = 0.001;
end pow;
```

**Comment by phannebohm** [comment]
> I think `pow` is just a red herring in this case, since as mentioned it already seems to be computed as exactly as possible.

Running this code
```C
#include <stdio.h>
#include <math.h>

int main(){
    printf("%20.20g\n", pow(10, -3));
    printf("%20.20g\n", 0.001);
    return 0;
}
```
with MinGW clang 16.0.0 (or older) outputs:
```
0.0010000000000000002377
0.0010000000000000000208
```
while with MinGW clang 16.0.2 (or all other compilers I tried) it outputs:
```
0.0010000000000000000208
0.0010000000000000000208
```

**Comment by casella** [comment]
OK, so I guess we have it now: some older compilers compute `pow(10, -3) = 0.0010000000000000002377` (15 zeros between `1` and `2`). That's a fairly good approximation of 0.001, but not good enough for Ryu to conclude that the smaller representation with the same bitwise representation is 0.001. Hence, Ryu will print `0.0010000000000000002` (15 zeros between `1` and `2`). I guess that's just one LSB more than `0.0010000000000000000208` (16 zeros between  `1` and `2`).

That was probably a compiler bug that was later fixed.

**Comment by casella** [comment]
> Instead of computing `pow(10,unit.prefixExpo.toReal())`, you could use a local lookup table if there are only a few possible exponents for `unit.prefixExpo`. That would solve the precision issue and might even be faster than calling `pow`.

That's probably the best idea.

**Comment by casella** [comment]
There it is:
```C
double pow_10(int x)
{
  static double pos[19] = {1, 1e1, 1e2, 1e3, 1e4, 1e5, 1e6, 1e7, 1e8, 1e9, 1e10, 1e11, 1e12, 1e13, 1e14, 1e15, 1e16, 1e17, 1e18};
  static double neg[19] = {1, 1e-1, 1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7, 1e-8, 1e-9, 1e-10, 1e-11, 1e-12, 1e-13, 1e-14, 1e-15, 1e-16, 1e-17, 1e-18};
  if (x > 18 || x < -18)
    return 0; // Should never get here, just to avoid buffer overflow
  if(x >= 0)
    return pos[x];
  else
    return neg[-x];
}
```

**Comment by adrpo** [comment]
omc on windows with msys2/ucrt:

```
adrpo33@ida-0030 UCRT64 /c/home/adrpo33/dev/OMTesting
$ omc convertUnits.mos
0.001
""
0.001
(true,0.001,0.0)
""
```

So merging #10939 and changing OMDev to the UCRT one would fix this issue.

**Comment by adeas31** [comment]
Fixed now after #10939
